static-metric/proc-macros/src/auto_flush_builder.rs

```diff
@@ -0,0 +1,816 @@
+                )*
+
+                #auto_flush_delegator
+                #outer_struct
+                #outer_impl
+            }
+        }
+    }
+
+    fn build_auto_flush_delegator(
+        metric: &MetricDef,
+        builder_contexts: &Vec<MetricBuilderContext>,
+    ) -> Tokens {
+        let inner_struct = &builder_contexts[0].inner_struct_name();
+        let last_builder_contexts = &builder_contexts
+            .last()
+            .expect("builder contexts should not be empty");
+        let last_delegator = last_builder_contexts.delegator_struct_name();
+        let metric_type = metric.metric_type.clone();
+
+        fn offset_fetcher(builder_context: &MetricBuilderContext) -> Tokens {
+            let struct_type = builder_context.inner_struct_name();
+            let struct_var_name = Ident::new(
+                struct_type.to_string().to_lowercase().as_str(),
+                Span::call_site(),
+            );
+
+            let member_type = builder_context.inner_member_type.clone();
+            let member_var_name = Ident::new(
+                member_type.to_string().to_lowercase().as_str(),
+                Span::call_site(),
+            );
+            let offset = Ident::new(
+                &format!("offset{}", builder_context.label_index + 1),
+                Span::call_site(),
+            );
+            let head = if builder_context.label_index == 0 {
+                quote! {
+                    let #struct_var_name = root_metric as *const #struct_type;
+                }
+            } else {
+                Tokens::new()
+            };
+
+            let body = quote! {
+                let #member_var_name = (#struct_var_name as usize + self.#offset) as *const #member_type;
+            };
+
+            let tail = if builder_context.is_last_label {
+                quote! {
+                    &*#member_var_name
+                }
+            } else {
+                Tokens::new()
+            };
+
+            quote! {
+                #head
+                #body
+                #tail
+            }
+        }
+
+        let offset_fetchers = builder_contexts
+            .iter()
+            .map(|m| offset_fetcher(m))
+            .collect::<Vec<Tokens>>();
+
+        let delegator_tokens = if metric_type.to_string().contains("Counter") {
+            quote! {
+                CounterDelegator<#inner_struct, #metric_type>
+            }
+        } else {
+            quote! {
+                HistogramDelegator<#inner_struct>
+            }
+        };
+
+        quote! {
+            impl #delegator_tokens for #last_delegator {
+                fn get_root_metric(&self) -> &'static LocalKey<#inner_struct> {
+                    self.root
+                }
+                fn get_local<'a>(&self, root_metric: &'a #inner_struct) -> &'a #metric_type {
+                   unsafe {
+                    #(
+                      #offset_fetchers
+                    )*
+                   }
```

Comment:
weird identation... I think it's appropriate to write them in a single line, or keep 4 space indentation